src/raz_meta.rs

```diff
@@ -112,4 +119,67 @@ impl FirstLast for usize {
 	fn last() -> Self { usize::max_value() }
 }
 
-// TODO: HashMap
\ No newline at end of file
+/// Metadata for names in a raz tree.
+///
+/// Hash is implemented by a no-op, since the data here
+/// can be found elsewhere in the tree. This is not intended
+/// to be used outside of a raz.
+#[derive(Clone,Eq,PartialEq,Debug)]
+pub struct Names(pub HashMap<Name,()>);
+impl Hash for Names {
+	/// does nothing
+	fn hash<H:Hasher>(&self, _state: &mut H) {}
+}
+
+#[derive(Clone,Debug)]
+pub enum NameIndex {
+	FarLeft,
+	FarRight,
+	Name(Name),
+}
+
+impl From<Name> for NameIndex {
+	fn from(nm: Name) -> Self { NameIndex::Name(nm) }
+}
+
+impl FirstLast for NameIndex {
+	fn first() -> Self { NameIndex::FarLeft }
+	fn last() -> Self { NameIndex::FarRight }
+}
+
+impl<E> RazMeta<E> for Names {
+	type Index = NameIndex;
+
+	fn from_none() -> Self { Names(HashMap::new()) }
+	fn from_vec(_vec: &Vec<E>) -> Self { Names(HashMap::new()) }
+	fn from_meta(l: &Self, r: &Self, _lev: u32, n: Option<Name>) -> Self {
+		let mut h = HashMap::new();
+		for k in l.0.keys() { h.insert(k.clone(),()); }
+		for k in r.0.keys() { h.insert(k.clone(),()); }
+		match n {None=>{},Some(nm)=>{ h.insert(nm,()); }}
+		Names(h)
+	}
+	fn choose_side(l: &Self, r: &Self, index: &Self::Index) -> SideChoice<Self::Index> {
+		match *index {
+			NameIndex::FarLeft => SideChoice::Left(NameIndex::FarLeft),
+			NameIndex::FarRight => SideChoice::Right(NameIndex::FarRight),
+			NameIndex::Name(ref nm) => {
+				match (l.0.contains_key(nm),r.0.contains_key(nm)) {
+					(true,true) => SideChoice::Here,
```

Comment:
Should this `(true, true)` case actually be a dynamic error?
The name `nm` is not unique, and appears (at least) twice, correct?